# 魔术方法  

---

## 上下文管理  

---

文件IO操作可以对文件对象使用上下文管理,使用with...as语法.  
```py
with open('test') as f:
    pass
```
仿照上面的例子写一个自己的类,实现上下文管理.  

```py
# with open('test') as f:
#     pass

# 仿照上面的例子写一个自己的类,实现上下文管理

class Point:
    pass

with Point() as p: #AttributrError: __exit__
    pass
```
输出结果为:

```py
Traceback (most recent call last):
    with Point() as p: #AttributrError: __exit__
AttributeError: __enter__
```
提示属性错误,没有`__exit__`,看来需要这个属性.  
某些版本会实体没有`__enter__`  

---
## 上下文管理对象  
当一个对象同时实现了`__enter__()`和`_-exit__()`方法,它就属于上下文管理的对象.  
|方法|意义|
|:--|:----|
|`__enter__`|进入与此对象相关的上下文管理,如果存在该方法,with语法会把该方法的返回值作为绑定到as子句中指定的变量上  |
|`__exit__`|推出与对象相关的上下文  |

---
编写一个类似上下文管理的类  
```py
from time import sleep

class Point:
    def __init__(self):
        print('init _____')
        sleep(2)
        print('end init ')
    def __enter__(self):
        print('enter now ')

    def __exit__(self, exc_type, exc_val, exc_tb):
        print('now exit over')

with Point() as p:
    print('in with -------')
    sleep(2)
    print('with is over')
```
输出结果为:

```py
init _____
end init 
enter now 
in with -------
with is over
now exit over

```
结论:
    实例化对象的时候,并不会调用enter,进入with语句块调用`__enter__`方法,然后执行语句体,最后离开with语句快的时候,调用`__exit__`方法.  
    with可以开启一个运行环境,在执行前做一些准备工作,执行后做一些收尾工作,  
    注意: with并不开启一个新的作用域
总结:
    with操作的是with后面的实例对象与as后面的别名无关,
    上下文光力士enter和exit相关,谭恩都是和实力相符and,他们两个需要成对出现,缺一不可,  
    它们一进一出,有进必有出,与进入之后操作的结果如何并无关系,既是时报错了,exit亦可以顺利的执行退出.  

---
## 上下文管理的安全性  


```py
from time import sleep

class Point:
    def __init__(self):
        print('init _____')
        sleep(2)
        print('end init ')
    def __enter__(self):
        print('enter now ')

    def __exit__(self, exc_type, exc_val, exc_tb):
        print('now exit over')

with Point() as p:
    print('in with -------')
    raise Exception('error')
    sleep(2)
    print('with is over')
```
输出结果为:

```py
init _____
end init 
Traceback (most recent call last):
enter now 
in with -------

now exit over
    raise Exception('error')
Exception: error
```
从上面的例子可以看出,enter和exit无论上下文是如何处理的都会被执行,所以说**上下文管理是安全的**  

极端的例子  
调用sys.exit(),他会退出当前解释器.  
打开Python解释器,在里面敲入sys.exit(),窗口直接关闭了,也就是说碰到这一句,Python运行环境直接退出了.  

---

```py
from time import sleep
import sys

class Point:
    def __init__(self):
        print('init _____')
        sleep(2)
        print('end init ')
    def __enter__(self):
        print('enter now ')

    def __exit__(self, exc_type, exc_val, exc_tb):
        print('now exit over')

with Point() as p:
    print('in with -------')
    sys.exit(1)
    sleep(2)
    print('with is over')
```
输出结果为:

```py
init _____
end init 
enter now 
in with -------
now exit over
```
从上面的执行结果来看,依然执行了`__exit__`函数,哪怕是退出了Python运行环境了.  
这说明了**上下文管理是很安全的**  

---
## with语句

理解 as字句
```py
class Point:
    def __init__(self):
        print('init')

    def __enter__(self):
        print('enter')

    def __exit__(self, exc_type, exc_val, exc_tb):
        print('exit')


f = open('4.py')
with f as p:
    print(f)
    print(p)
    print(p is f)
    print(p == f)

print('*'*50)
p = Point()
with p as f:
    print('in with =========')
    print(p == f)
    print('with over')

print('end','---'*10)
```
输出结果为:

```py
<_io.TextIOWrapper name='4.py' mode='r' encoding='cp936'>
<_io.TextIOWrapper name='4.py' mode='r' encoding='cp936'>
True
True
**************************************************
init
enter
in with =========
False
with over
exit
end ------------------------------
```
问题在于`__enter__`方法上,他将自己的返回值赋给了f.修改上例  

---


enter返回实例自身;  
```py
class Point:
    def __init__(self):
        print('init')

    def __enter__(self):
        print('enter')
        return self #增加返回值

    def __exit__(self, exc_type, exc_val, exc_tb):
        print('exit')


f = open('4.py')
with f as p:
    print(f)
    print(p)
    print(p is f)
    print(p == f)

print('*'*50)
p = Point()
with p as f:
    print('in with =========')
    print(p == f)
    print('with over')

print('end','---'*10)
```
输出结果为:

```py
<_io.TextIOWrapper name='4.py' mode='r' encoding='cp936'>
<_io.TextIOWrapper name='4.py' mode='r' encoding='cp936'>
True
True
**************************************************
init
enter
in with =========
True
with over
exit
end ------------------------------
```
with 语法,会调用with后的对象`__enter__`方法,如果有as,则将该方法的返回值赋给as字句的变量.  
上例,可以等价为`f = p.__enter__()`  
`p.__enter__()`返回的有事self,就是实例本身,既是p本身因此也可以写为f = p  

---
## 方法的参考  
`__enter__`方法 没有其他的参数  
`__exit__`方法有三个参数  
`__exit__(self, exc_type, exc_val, exc_tb)`  
这三个参数都是与异常有关的.  
如果该上下文退出时没有异常,那么这三个参数都是None  
如果有异常,参数的意义如下:  
`exc_type`,异常类型  
`exc_val`,异常的值 
`exc_tb`,异常的追踪信息,    

实现代码为:  
```py
class Point:
    def __init__(self):
        print('init')

    def __enter__(self):
        print('enter')
    def __exit__(self, exc_type, exc_val, exc_tb):
        print(1,exc_type)
        print(2,exc_val)
        print(3,exc_tb)
        print('exit end ---=--')

p = Point()
with p as f:
    print('in with ---------')
    print(1/0)
    print('with end')

print('='*30)
```
输出结果为:

```py
Traceback (most recent call last):
init

enter
    print(1/0)
in with ---------
ZeroDivisionError: division by zero
1 <class 'ZeroDivisionError'>
2 division by zero
3 <traceback object at 0x00D41D00>
exit end ---=--

```
---
##  上下文管理练习  
位加法函数计时  
方法1: 使用装饰器显示该函数的执行时长  
方法2: 使用上下文管理来显示该函数的执行时长,  
---


方法1:装饰器实现
```py
from time import sleep
from datetime import datetime

def timeit(fn):
    def wapper(*args):
        start = datetime.now()
        ret = fn(*args)
        print((datetime.now() - start).total_seconds())
        return ret
    return wapper

@timeit #add_new = timeit(add_old)  add_old的参数是传到了 timeit.wapper里面
def add(x=5,y=6):
    sleep(2)
    return x+y

print(add(2,3))
```
输出结果为:

```py
2.000263
5
```
---
方法二:上下文实现
分析:可使用,init记录需要计时的函数,enter记录进入类的时间,并执返回函数对象,在上下文操作的时候执行函数,等到函数执行完毕要退出的时候,使用exit来计算出时间.  但是这里记录的只能是总时间,却不能像是装饰器一样的记录单个的时间.  
```py
from time import sleep
from datetime import datetime

class Timeit:
    def __init__(self,fn):
        self.fn = fn

    def __enter__(self):
        self.start = datetime.now()
        return self.fn

    def __exit__(self, exc_type, exc_val, exc_tb):
        print((datetime.now() - self.start).total_seconds())

def add(x=5,y=6):
    sleep(2)
    return x+y

with Timeit(add):
    print(add(2,3))
```
输出结果为:

```py
5
2.000007
```
---

装饰器与上下文一并作用于函数的时候,既能记录单个函数的执行时间,也能记录整个上下文的执行时间.
```py
from time import sleep
from datetime import datetime

class Timeit:
    def __init__(self,fn):
        self.fn = fn

    def __enter__(self):
        self.start = datetime.now()
        return self.fn

    def __exit__(self, exc_type, exc_val, exc_tb):
        print((datetime.now() - self.start).total_seconds())

def timeit(fn):
    def wapper(*args):
        start = datetime.now()
        ret = fn(*args)
        print((datetime.now() - start).total_seconds())
        return ret
    return wapper

@timeit #add_new = timeit(add_old)  add_old的参数是传到了 timeit.wapper里面
def add(x=5,y=6):
    sleep(2)
    return x+y


with Timeit(add):
    print(add(2,3))
    print(add())
```
输出结果为:

```py
2.000215
5
2.000004
11
4.000219
```
---

另一种方法,就是把类实例变成可调用对象来实现
```py
from time import sleep
from datetime import datetime

class Timeit:
    def __init__(self,fn):
        self.fn = fn

    def __enter__(self):
        self.start = datetime.now()
        return self.fn

    def __exit__(self, exc_type, exc_val, exc_tb):
        print((datetime.now() - self.start).total_seconds())

    def __call__(self, x,y):
        return self.fn(x,y)

def timeit(fn):
    def wapper(*args):
        start = datetime.now()
        ret = fn(*args)
        print((datetime.now() - start).total_seconds())
        return ret
    return wapper

@timeit #add_new = timeit(add_old)  add_old的参数是传到了 timeit.wapper里面
def add(x=5,y=6):
    sleep(2)
    return x+y


with Timeit(add) as a:
    print(a(2,3))
    print(a())

```
输出结果为:

```py
2.000251
5
2.000179
11
4.00043
```
---

把类当多装饰器来实现
```py
from time import sleep
from datetime import datetime

class Timeit:
    def __init__(self,fn):
        self.fn = fn

    def __enter__(self):
        self.start = datetime.now()
        return self.fn

    def __exit__(self, exc_type, exc_val, exc_tb):
        print((datetime.now() - self.start).total_seconds())

    def __call__(self,*args,**kwargs):
        start = datetime.now()
        ret = self.fn(*args,**kwargs)
        print((datetime.now() - start).total_seconds())
        return ret

@Timeit
def add(x=5,y=6):
    sleep(2)
    return x+y

print(a(2,3))
print(a(5,11))
print(a())
```
输出结果为:

```py
2.000695
5
2.0002
16
2.000507
11

```
---
类装饰和装饰器属性的拷贝

```py
from time import sleep
from datetime import datetime
from functools import wraps

class Timeit:
    '''this is Timeit class'''
    def __init__(self,fn):
        self.fn = fn
        #把函数对象的文档字符串赋给类
        #self.__doc__ = fn.__doc__
        # update_wrapper(self,fn)
        wraps(fn)(self)#这一句和上面的依据作用是一样的
        #这样就实现了属性的拷贝了,
        #这个不同于函数的装饰器,函数装饰器直接@wraps就OK了,类里面你保护知道该怎么覆盖,就只能把wraps当做函数来覆盖了

    def __enter__(self):
        self.start = datetime.now()
        return self.fn

    def __exit__(self, exc_type, exc_val, exc_tb):
        print((datetime.now() - self.start).total_seconds())

    def __call__(self,*args,**kwargs):
        start = datetime.now()
        ret = self.fn(*args,**kwargs)
        print((datetime.now() - start).total_seconds())
        return ret

@Timeit
def add(x=5,y=6):
    '''this is add function'''
    sleep(2)
    return x+y

print(add(2,3))
print(add.__doc__)
print(Timeit(add).__doc__)
```
输出结果为:

```py
2.000172
5
this is add function
this is add function
```
上面的类既可作上下文管理,又可以做装饰器  

---

## 上下文应用场景 
---
* 
    - 1.增强代码功能  
        在代码执行的前后增加代码,以增强其功能,类似装饰器的功能.  
    - 2.资源打开   
        打开了资源需要关闭,例如文件对象,网络连接,数据库连接等.
    - 3.权限验证  
        执行代码之前,做权限的验证,在`__enter__`和`__exit__`中处理.  

---
## contextlib.contextmanager  
---
contextlib.contextmanage  
他是一个装饰器实现上下文管理,装饰一个**函数**,而不是用像类一样实现`__enter__`和`__exit__`方法.  
对下面的函数有要求: 必须有yield,也就是这个函数必须返回一个生成器,且只有yired一个值.  

也就是装饰器接受一个生成器对象作为参数.  

```py
from contextlib import contextmanager

@contextmanager
def foo():
    print('enter') #yield之上相当于__enter__()
    yield 'yield'#yield 5 yield的值只能有一个,作为__enter__方法的返回值
    print('exit') #yield之下相当于__exit__()

with foo() as f:
    print(f)
```
输出结果为:

```py
enter
yield
exit
```
f接受yield语句的返回值  
上面的程序看似不错,增加一个异常试试,发现不能保证exit的执行怎么办?

解决办法 增加try finally  
这样就能解决exit不能执行的问题了
---

```py
from contextlib import contextmanager

@contextmanager
def foo():
    print('enter') #yield之上相当于__enter__()
    try:
        yield 'yield'#yield 5 yield的值只能有一个,作为__enter__方法的返回值
    finally:
        print('exit') #yield之下相当于__exit__()

with foo() as f:
    raise Exception()
    print(f)
```
输出结果为:

```py
enter
Traceback (most recent call last):
exit
    raise Exception()
Exception
```
* 上例这么做又有什么意义呢?  
    - 把yield之前当做`__enter__`方法执行  
    - 把yield之后当做`__exit__`方法执行  
    - 把yield的值作为`__enter__`的返回值  

---


```py
from contextlib import contextmanager
from datetime import datetime
from time import sleep

@contextmanager
def add(x,y):
    start = datetime.now() #yield之上相当于__enter__()
    try:
        sleep(2)
        yield x+y#yield 5 yield的值只能有一个,作为__enter__方法的返回值
    finally:
        print ((datetime.now() - start).total_seconds())

with add(4,5) as f:
    print(f)
```
输出结果为:

```py
9
2.000165

```
总结:  
    如果业务逻辑简单可以使用函数contextlib.contextmanage装饰器方式,如果业务复杂,用类方式`__enter__`和`__exit__`方法更方便.  